File: src/agents/classifier.py
import os
import json
import httpx
from tenacity import retry, stop_after_attempt, wait_exponential
from src.utils.logger import logger

OPENROUTER_URL = "https://openrouter.ai/api/v1/chat/completions"
MODEL = os.getenv("OPENROUTER_MODEL", "openrouter/auto")
TIMEOUT = int(os.getenv("LLM_TIMEOUT", "30"))

CATEGORIES = [
    "infectious_disease", "outbreak", "pandemic", "vaccine",
    "pharmaceutical", "clinical_research", "public_health",
    "mental_health", "chronic_disease", "medical_technology", "other",
]

RISK_CRITERIA = """
- critical: Active outbreak, pandemic, bioterrorism, mass casualties, emergency alert
- high: New pathogen, rapid spread, drug-resistant strain, significant mortality
- medium: Disease monitoring, new research findings, regional health concerns
- low: General health news, routine updates, background research
"""

SYSTEM_PROMPT = f"""You are a medical risk assessment system. Classify the article and return JSON with:
- risk_level: one of "critical", "high", "medium", "low"
- category: one of {CATEGORIES}
- risk_reason: one sentence explaining the risk level

Risk level criteria:
{RISK_CRITERIA}

Return only valid JSON."""


@retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=2, max=10))
async def classify(title: str, summary: str, diseases: list[str]) -> dict:
    """위험도와 카테고리를 분류."""
    api_key = os.environ.get("OPENROUTER_API_KEY", "")
    if not api_key:
        return _default_classification()

    user_msg = f"""Title: {title}
Summary: {summary}
Diseases mentioned: {", ".join(diseases) if diseases else "None"}

Classify this article."""

    try:
        async with httpx.AsyncClient(timeout=TIMEOUT) as client:
            resp = await client.post(
                OPENROUTER_URL,
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                    "HTTP-Referer": "https://github.com/medical-news-agent",
                    "X-Title": "Medical News Agent",
                },
                json={
                    "model": MODEL,
                    "messages": [
                        {"role": "system", "content": SYSTEM_PROMPT},
                        {"role": "user", "content": user_msg},
                    ],
                    "temperature": 0.1,
                    "max_tokens": 200,
                    "response_format": {"type": "json_object"},
                },
            )
            resp.raise_for_status()
            raw = resp.json()["choices"][0]["message"]["content"]
            result = json.loads(raw)
            return {
                "risk_level": result.get("risk_level", "unknown"),
                "category": result.get("category", "other"),
                "risk_reason": result.get("risk_reason", ""),
            }
    except Exception as e:
        logger.error(f"[Classifier] 분류 오류: {e}")
        raise


def _default_classification() -> dict:
    return {"risk_level": "unknown", "category": "other", "risk_reason": ""}

File: src/agents/classifier.ts
import { chat, parseJsonSafe } from '@/lib/openrouter';
import { logger } from '@/lib/logger';
import type { ClassificationResult, RiskLevel, ArticleCategory } from '@/types';
import { ARTICLE_CATEGORIES, RISK_LEVELS } from '@/types';

const SYSTEM_PROMPT = `You are a medical news risk and category classifier.

## Categories (choose exactly one Korean label):
- 감염병: Infectious diseases, outbreaks, pandemics, epidemics, bacteria, viruses, parasites, zoonosis
- 만성질환: Chronic diseases, cancer, diabetes, cardiovascular, neurodegenerative, obesity
- 백신: Vaccines, vaccination programs, immunization, vaccine safety/efficacy
- 제약/신약: New drugs, drug approvals, clinical trials, FDA/EMA, pharmaceutical
- 공공보건: Public health policy, epidemiology, health statistics, prevention, health education
- AI 의료: AI in medicine, machine learning diagnostics, digital health, health tech, wearables
- 병원/의료정책: Hospitals, healthcare systems, insurance, medical workforce, health regulations
- 기타: Any medical news that does not fit the above categories

## Risk levels:
- critical: Active outbreak, declared pandemic/emergency, bioterrorism, mass casualties
- high: Novel pathogen identified, rapid multi-country spread, drug-resistant strain, high mortality event
- medium: Emerging health concern, new research findings, regional outbreak under control
- low: Routine health news, general research updates, informational articles

Return a JSON object with exactly:
{ "risk_level": "<one of: critical|high|medium|low>", "category": "<Korean label>", "risk_reason": "<one sentence>" }

No extra text. No markdown.`;

function buildPrompt(title: string, summary: string, diseases: string[]): string {
  return `Title: ${title}
Summary: ${summary || '(none)'}
Diseases/Pathogens: ${diseases.length > 0 ? diseases.join(', ') : 'none'}

Classify this article.`;
}

function isValidRiskLevel(v: unknown): v is RiskLevel {
  return typeof v === 'string' && (RISK_LEVELS as string[]).includes(v);
}

function isValidCategory(v: unknown): v is ArticleCategory {
  return typeof v === 'string' && (ARTICLE_CATEGORIES as string[]).includes(v);
}

export async function classify(
  title: string,
  summary: string,
  diseases: string[]
): Promise<ClassificationResult> {
  const fallback: ClassificationResult = {
    risk_level: 'unknown',
    category: '기타',
    risk_reason: '',
  };

  if (!process.env.OPENROUTER_API_KEY) return fallback;

  let lastError: unknown;
  for (let attempt = 1; attempt <= 3; attempt++) {
    try {
      const raw = await chat(
        [
          { role: 'system', content: SYSTEM_PROMPT },
          { role: 'user', content: buildPrompt(title, summary, diseases) },
        ],
        { temperature: 0.1, maxTokens: 200, jsonMode: true }
      );

      const parsed = parseJsonSafe<Record<string, unknown>>(raw);
      if (!parsed) {
        logger.warn(`[Classifier] JSON 파싱 실패 (시도 ${attempt})`);
        continue;
      }

      return {
        risk_level: isValidRiskLevel(parsed.risk_level) ? parsed.risk_level : 'unknown',
        category: isValidCategory(parsed.category) ? parsed.category : '기타',
        risk_reason: typeof parsed.risk_reason === 'string' ? parsed.risk_reason : '',
      };
    } catch (e) {
      lastError = e;
      logger.warn(`[Classifier] LLM 오류 (시도 ${attempt}): ${e}`);
      if (attempt < 3) await sleep(attempt * 2000);
    }
  }

  logger.error(`[Classifier] 3회 실패: ${lastError}`);
  return fallback;
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
